llm/tests/compare_hf_v_mosaic_gpt.py

```diff
@@ -0,0 +1,140 @@
+import warnings
+
+import torch
+from omegaconf import OmegaConf as om
+from composer.utils import reproducibility
+
+from main import build_composer_model
+
+
+def test_compare_hf_v_mosaic_gpt():
+    warnings.filterwarnings(action='ignore', message='Torchmetrics v0.9 introduced a new argument class property')
+    device = 'cuda'
+
+    # get 125m config
+    with open('yamls/mosaic_gpt/125m.yaml') as f:
+        cfg = om.load(f)
+
+    # modify cfg for HF GPT2 compatibility
+    cfg.max_seq_len = 1024  # GPT2 seq len (GPT3 uses 2048)
+    cfg.model.max_seq_len = cfg.max_seq_len
+    cfg.train_loader.max_seq_len = cfg.max_seq_len
+    cfg.eval_loader.max_seq_len = cfg.max_seq_len
+    cfg.model.device = device
+    cfg.precision = 'fp16'
+    # set dropout prob
+    cfg.model.resid_pdrop = 0.1
+    cfg.model.emb_pdrop = 0.1
+    # attn_dropout is imtegrated into the FlashMHA kernel
+    # given this, it will generate different drop idx when compared to nn.Dropout
+    # reguradless of if rng is seeded.
+    cfg.model.attn_pdrop = 0.0
+
+    # get equivalent config for HF model
+    hf_cfg = om.create({'name': 'hf_causal_lm', 'hf_config_name_or_path': 'gpt2'})
+
+    # set seed
+    reproducibility.seed_all(cfg.seed)
+
+    # Build Model
+    print('Initializing model...')
+
+    print(cfg)
+    model = build_composer_model(cfg.model)
+    n_params = sum(p.numel() for p in model.parameters())
+
+    hf_model = build_composer_model(hf_cfg).to(device)
+    hf_n_params = sum(p.numel() for p in hf_model.parameters())
+
+    assert hf_n_params == n_params
+
+    # Get batch size info
+    batch_size = 2
+    
+    # generate random input branch
+    batch = {}
+    batch['input_ids']      = torch.randint(low=0, high=cfg.model.vocab_size, size=(batch_size, cfg.max_seq_len)).to(device)
+    batch['attention_mask'] = torch.ones(size=(batch_size, cfg.max_seq_len), dtype=torch.int64).to(device)
+    batch['labels']         = torch.randint(low=0, high=cfg.model.vocab_size, size=(batch_size, cfg.max_seq_len)).to(device)
+
+    # # set dropout prob to 0 for potentially more numerically precise comparisons
+    # model.model.transformer.emb_drop.p = 0.0
+    # hf_model.model.transformer.drop.p = 0.0
+
+    # for b in model.model.transformer.blocks:    b.resid_mlp_dropout.p = 0.0
+    # for b in hf_model.model.transformer.h:      b.mlp.dropout.p = 0.0
+
+    # for b in model.model.transformer.blocks:    b.resid_attn_dropout.p = 0.0
+    # for b in hf_model.model.transformer.h:      b.attn.resid_dropout.p = 0.0
+
+    # # attn_dropout is imtegrated into the FlashMHA kernel and must therefore be set to 0
+    # for b in model.model.transformer.blocks:    b.attn_dropout.p = 0.0 # set in cfg
+    for b in hf_model.model.transformer.h:      b.attn.attn_dropout.p = 0.0
+
+    hf_model.train()
+    model.train()
+
+    # UTIL: can be used to verify that models are not the same at init
+    # with torch.autocast(device_type='cuda', dtype=torch.float16):
```

Comment:
This seems valuable to keep, and we can add an assert that the outputs are not equal?

Reply:
There exists a universe, where by chance, the outputs, at init, are the same.
I'll add a warnings.warn if that is the case ie
```
    if hf_model_fwd.mean().allclose(model_fwd.mean()):
        warn_msg = f'WARNING: model_fwd ({model_fwd}) and hf_model_fwd ({hf_model_fwd}) are very close at init.'
        raise warnings.warn(warn_msg)
```